pageserver/src/consumption_metrics.rs

```diff
@@ -261,7 +291,7 @@ pub async fn collect_metrics_task(
     });
 
     if current_metrics.is_empty() {
-        trace!("no new metrics to send");
+        info!("no new metrics to send");
```

Comment:
> It is not explained why it was decided to calculate logical size in background. Before it is calculated incrementally. May be I missed something but logical=synthetic size is one we are limiting for free tier and what customers are paying for. If so, is it acceptable to calculate it once per 10 minutes? So I do some massive insert, what to check used size but see... old value. Once again sorry if it is explained somewhere and I just missed it.

@knizhnik 

1) logical != synthetic.
logical size is per-timeline (see TimelineLogicalSize)
This is what `neon.pg_cluster_size()` returns.
IIUC, we're limiting usage based on this one.

synthetic size is a... well, synthetic per-tenant size that we compute to understand tenant's size, taking into account branching structure, COW and gc_cutoff.
 
This is what `/v1/tenant/:tenant_id/size` endpoint returns.
And we haven't really used it before, except for testing. That's why bug was unnoticed for a while.

2) Why separate background worker?

Because computation can be quite heavy operation (we need to gather info about all branches and calculate their sizes at branchpoint lsns) and we want to be able to tweak it independently of other processes. 

You can find some details in [Consumption metrics RFC](https://github.com/neondatabase/neon/pull/2884)

@koivunej, @hlinnaka , please correct me, if I am wrong

Reply:
Thank you for clarification.
Sorry, I have not read this RFC before.
But still I have some questions.
So do I correctly understand that **synthetic** size is physical size of all tenant layer files both at loaded (on pageserver) and offloaded (in S3)?
And **real storage size** is total size of loaded tenant's layers in pageserver?
Both are calculated per-tenant, doesn't mater how much branches this tenant has, right?

In this case the most efficient and at the same time easiest way to calculate both values is to traverse layer maps in all tenants timelines. Hierarchy of timeline is not important. In case of **real size** we should just look at loaded layers,
for synthetic - consider all layers.

The I wonder why this sophisticated logic with sorting updates `sort_updates_in_tree_order` is needed and why do we need to take in account branch hierarchy? Or do we want to exclude from **synthetic size" layers which are out of PiTR region but for some reasons are not yet collected by GC and so stored in pageserver or in S3? Is it really important for us?

Reply:
The synthetic size is not just a sum of all layer sizes, it is more like "let's sum **logical** sizes of all tenant's timeline minus intersecting COW segments"
I haven't found document describing tenant sizing model. @koivunej , is there any RFC/doc?

For more details about current implementation, you can check `gather_inputs` and `calculate` functions in size.rs.
`calculate` logic expects that we handle branches in correct order - only handle branch, when it's parent have been processed)

In a rare(?) case  when two branches have branched at the same LSN, they could have been sorted incorrectly.
i.e. for operations
branch a from main at lsn 10
branch b from a at lsn 10

we could process b before a and fail, because a is unknown branch.

That's what I hit with my test. See https://github.com/neondatabase/neon/issues/3179
This PR fixes it. Maybe code can be simpler, though. I've asked @koivunej to review.

P.S. To add to the confusion,  there is also a draft PR https://github.com/neondatabase/neon/pull/2817 that reworks this model )

Reply:
> I haven't found document describing tenant sizing model. <ping>, is there any RFC/doc?

I don't think so. The model itself is the best documentation, which is further "inlined" #draft. Heikki did provide a motivating example behind in https://github.com/neondatabase/neon/pull/2884#discussion_r1029365278 which I think is very good, but the algorithm that follows is not described publicly at least, the [slack thread](https://neondb.slack.com/archives/C043L4SG17B/p1666089915343109?thread_ts=1665597859.715069&cid=C043L4SG17B) where I ask a lot questions might have a better description.

Also note the #2817 rewrite, which as far as I know, didn't aim to change many things, but shuffles things around, inlines, perhaps aimed to make the algorithmic implementation better.

Reply:
> Is it intentional change? Do we really need to log this message each time when no new metric is found?

nope, I have a TODO about this
I'll fix it after main part of the code is polished

Reply:
1. why do we ever need such "synthetic" size?
AFAIK, the idea is to have some objective metric on how much data user really stores in neon.
Physical size depends on our implementation details (i.e. how often we materialize pages, do we use compression, etc).
So it is not very transparent parameter for billing purposes.

synthetic size answers the question "How much space my tables would consume in postgres with copy-on-write"
Very roughly speaking for branch main (10 Gb) + branch a (extra 1Gb) forked from main 
synthetic size is 11GB.

2. Do we report it to users?
AFAIK, we're going to show it on billing page.

3. Do we enforce some constraints based on this size?
I am not sure if we have per-tenant size limits or just per-timeline. @acervantes23 , can you please help?

4. 10minutes is ok. I had the same question before. We charge users on per hour basis, so this level of granularity is enough. Anyway, there is a setting, so we can tweak it later.

@hlinnaka , please, correct me, if I am wrong. I want to write some internal doc based on this thread.

Reply:
Frankly speaking this idea of synthetic  size is unclear to me. First of all how we are going to use this metric? Do we plan to charge clients for this size? Or limit size of client database? In both cases 10 minutes interval of recalculation of this metrics seems to be unacceptable.

Sorry, I am not familiar with our pricing policy. Will we charge users for logical size of his database doesn't mater where layer files are located - at S3 or at pageserver?

Reply:
I resolve this conversation, because it's out of scope here.
I'll write a doc about our metrics in a few days and share it with you.

Reply:
@lubennikovaav following up on your question:

> I am not sure if we have per-tenant size limits or just per-timeline. @acervantes23 , can you please help?

We may impose some limitations for the free tier based on synthetic size (still working that out).  In that case the limitations would be applied at the tenant level.

Let me know if that doesn't answer your question.

Cc @stepashka